openlibrary/plugins/worksearch/code.py

```diff
@@ -202,64 +205,73 @@ def run_solr_query(param = {}, rows=100, page=1, sort=None, spellcheck_count=Non
-    fl = ','.join(fields)
-    solr_select = solr_select_url + "?fq=type:work&q.op=AND&start=%d&rows=%d&fl=%s" % (offset, rows, fl)
     if q_list:
         if use_dismax:
-            q = web.urlquote(' '.join(q_list))
-            solr_select += "&defType=dismax&qf=text+title^5+author_name^5&bf=sqrt(edition_count)^10"
+            params.append(('q', ' '.join(q_list)))
+            params.append(('defType', 'dismax'))
+            params.append(('qf', 'text title^5 author_name^5'))
+            params.append(('bf', 'sqrt(edition_count)^10'))
         else:
-            q = web.urlquote(' '.join(q_list + ['_val_:"sqrt(edition_count)"^10']))
-        solr_select += "&q=%s" % q
-    solr_select += '&spellcheck=true&spellcheck.count=%d' % spellcheck_count
-    solr_select += "&facet=true&" + '&'.join("facet.field=" + f for f in facet_fields)
+            params.append(('q', ' '.join(q_list + ['_val_:"sqrt(edition_count)"^10'])))
 
     if 'public_scan' in param:
         v = param.pop('public_scan').lower()
         if v in ('true', 'false'):
             if v == 'false':
                 # also constrain on print disabled since the index may not be in sync
                 param.setdefault('print_disabled', 'false')
-            solr_select += '&fq=public_scan_b:%s' % v
+            params.append(('fq', 'public_scan_b:%s' % v))
 
     if 'print_disabled' in param:
         v = param.pop('print_disabled').lower()
         if v in ('true', 'false'):
-            solr_select += '&fq=%ssubject_key:protected_daisy' % ('-' if v == 'false' else '')
+            minus = '-' if v == 'false' else ''
+            params.append(('fq', '%ssubject_key:protected_daisy' % minus))
 
-    k = 'has_fulltext'
-    if k in param:
-        v = param[k].lower()
+    if 'has_fulltext' in param:
+        v = param['has_fulltext'].lower()
         if v not in ('true', 'false'):
-            del param[k]
-        param[k] == v
-        solr_select += '&fq=%s:%s' % (k, v)
-
-    for k in facet_list_fields:
-        if k == 'author_facet':
-            k = 'author_key'
-        if k not in param:
+            del param['has_fulltext']
+        params.append(('fq', 'has_fulltext:%s' % v))
+
+    for field in facet_list_fields:
+        if field == 'author_facet':
+            field = 'author_key'
+        if field not in param:
             continue
-        v = param[k]
-        solr_select += ''.join('&fq=%s:"%s"' % (k, url_quote(l)) for l in v if l)
-    if sort:
-        solr_select += "&sort=" + url_quote(sort)
+        values = param[field]
+        params += [('fq', '%s:"%s"' % (field, val)) for val in values if val]
 
-    solr_select += '&wt=' + url_quote(param.get('wt', 'standard'))
+    if sort:
+        params.append(('sort', sort))
 
-    solr_result = execute_solr_query(solr_select)
+    params.append(('wt', param.get('wt', 'standard')))
+    params = [(k, v.encode('utf-8') if isinstance(v, six.string_types) else v)
+              for (k, v) in params]
```

Comment:
@cdrini I've figured out what this is doing now -- 
`urlencode()` will error if it gets a bare unicode parameter value, e.g.:
`urlencode({'key': u'αβγ'})`
=> `UnicodeEncodeError: 'ascii' codec can't encode characters in position 0-2: ordinal not in range(128)`

so a fix is to
`urlencode({'key': u'αβγ'.encode('utf8')})`
=> `'key=%CE%B1%CE%B2%CE%B3' `

The cause of all the issues is that the old `urlib` doesn't handle Unicode at all. `requests` will, and make this a lot clearer.

I have no idea why unicode inside lists _is_ correctly encoded by `urllib`:
`urlencode({'key': [u'αβγ']})`
=> `'key=%5Bu%27%5Cu03b1%5Cu03b2%5Cu03b3%27%5D'`

The line `(k, v.encode('utf-8') if isinstance(v, six.string_types) else v)` will leave everything that is not a string alone, so `int`s, which _are_ in `params` and lists (no idea if they are to be expected), will be passed to `urlencode()` unaltered, which `urlencode` will handle correctly ............. sooooo I think this is doing the right thing in most case, it's just a pain having to read out-of-date docs to confirm old behaviour.

I assume all of the params are unicode strings, because

`urlencode({'key': 'αβγ'})`
and
`urlencode({'key': b'αβγ'})`
give the expected encoding:
`'key=%CE%B1%CE%B2%CE%B3'`

but either of those already encoded data are passed to `encode('utf-8')` will error, so if `v = b' αβγ'` the above code will throw an exception. I'm over analysing, but I forgot how awkward dealing with `urllib` is, and was worried about double encoding. 

**EDIT** ... I just looked more into the various `six` types ...
`six.text_type` will give better protection:

Current:
`[v.encode('utf-8') if isinstance(v, six.string_types) else v for v in [1, ['test'], 'αβγ', b'αβγ', u'αβγ'.encode('utf8'), u'αβγ']]`
=> `UnicodeDecodeError: 'ascii' codec can't decode byte 0xce in position 0: ordinal not in range(128)`

Proposed:
`[v.encode('utf-8') if isinstance(v, six.text_type) else v for v in [1, ['test'], 'αβγ', b'αβγ', u'αβγ'.encode('utf8'), u'αβγ']]`
=> `[1, ['test'], '\xce\xb1\xce\xb2\xce\xb3', '\xce\xb1\xce\xb2\xce\xb3', '\xce\xb1\xce\xb2\xce\xb3', '\xce\xb1\xce\xb2\xce\xb3']`

and with `urlencode()` thrown into the mix:
`[urlencode({'k': v.encode('utf-8') if isinstance(v, six.text_type) else v}) for v in [1, ['test'], 'αβγ', b'αβγ', u'αβγ'.encode('utf8'), u'αβγ']]`
=> `['k=1', 'k=%5B%27test%27%5D', 'k=%CE%B1%CE%B2%CE%B3', 'k=%CE%B1%CE%B2%CE%B3', 'k=%CE%B1%CE%B2%CE%B3', 'k=%CE%B1%CE%B2%CE%B3']`

Reply:
@cdrini This is the review I was in the middle of writing and testing as requested when this got merged.

Reply:
Ahhh, that's my bad 😖 I thought you were the assignee which is why I asked for the review, but thought you were offline. Proposal seems 👍 ; I'll open a new PR and get that in before tomorrow's deploy.

Reply:
Sorry for the delay! I tried switching to `text_type` a few days ago, but it was causing errors for some reason :/ Man, this is string encoding is _easily_ one of the most frustrating things about using python :P